src/sipmpdf/functions.py

```diff
@@ -216,47 +245,384 @@ def darkcurrent_response(x: np.float64,
   Returns
   -------
   np.float64
-      Probability density of observing a dark current response of
+      Probability density of observing a dark current response
   """
-  kern = kernel_switch(x, pedestal, gain, resolution)
+  kern = kernel_switch(x, gain, resolution)
+
+  eps = gain * resolution
+
+  left = 1 / x
+  right = 1 / (gain - x)
+  norm = 2 * (kern.log(1 - resolution) - kern.log(resolution))
+  return kern.where(((x > eps) & (x < (gain - eps))),  #
+                    (left + right) / norm, 0)
+
+
+def darkcurrent_response(x: np.float64,
+                         gain: np.float64,
+                         resolution: np.float64 = 1e-4) -> np.float64:
+  """
+  Modified dark current response, as the unmodified dark current response has
+  excessively sharp features that can make other numerical tools unstable, the
+  modified version attempts smooth out the dark current response by substituting
+  out the sharp features at x~pedestal and x~(pedestal+gain) with a normal
+  response with a width of ~0.01 resolution (since we don't expect any detector
+  to have such low a noise, 0.01 feels like a reasonable cutoff)
+
+  Here we modify the function to preserve the overall normalization and the 0-th
+  order smoothness of the function (function is continuous, but derivative is
+  not):
+
+  - If x is in the "nominal region" of (x > pedestal + 0.01 * gain) and (x <
+    pedestal - 0.01 * gain), we will return the original response function.
+
+  
+  Parameters
+  ----------
+  x : np.float64
+      Observation
+  gain : np.float64
+      The gain of the SiPM
+  resolution : np.float64, optional
+      resolution factor, relative to the SiPM gain factor, by default 1e-4
 
+  Returns
+  -------
+  np.float64
+      Smoothed out probability density of observing a dark current response
+  
+  """
+  kern = kernel_switch(x, gain, resolution)
+  coarse_res = resolution * 50
+  coarse_res = kern.where(coarse_res > kern.tofloat64(0.01),
+                          kern.tofloat64(0.01), coarse_res)
   eps = gain * resolution
-  lo = pedestal
-  up = pedestal + gain
-  return kern.where((x > (lo + eps)) & (x < (up - eps)),
-                    ((1 / (x - lo)) + (1 / (up - x))) / (2 * kern.log(
-                      (up - lo - eps) / eps)), 0)
+  coarse = gain * coarse_res
+
+  nominal = kern.where(((x > coarse) & (x < (gain - coarse))),  #
+                       darkcurrent_response_original(x, gain, resolution), 0.0)
+
+  outer = normal(x, 0, coarse) + normal(x, gain, coarse)
+  outer_norm = (kern.log(coarse_res) - kern.log(resolution)) / 2.0 / (
+    kern.log(1 - resolution) - kern.log(resolution))
+  return nominal + outer * outer_norm
 
 
 def ap_response_smeared(x: np.float64, smear: np.float64, n_ap: np.int64,
                         beta: np.float64) -> np.float64:
-  kern = kernel_switch(x, smear, n_np, beta)
-  pass
+  """
+  Afterpulse response given a fixed number of after pulses occurs accounting for the effects of random noise.
```

Comment:
"Afterpulse response, assuming a fixed number of afterpulses, taking into account random noise effects."